### ISSUE TITLE
If system command call QuPath, Bioformats Extension was not usable

### ISSUE THREAD

**Issue by rmd06** [qupath/qupath#260]
Problem of run script from system command when image was opened using BioFormats Extension

I want to using QuPath open an image and run a groovy script by command like this:

 "C:\Program Files\QuPath\app\QuPathApp.jar"  -image "D:\QMDownload\6\[failed_handOK]Leica_scn\Leica-Fluorescence-1.scn"  -script "D:\\QMDownload\\6\\Tiling_a_whole_slide_image.groovy"

If the image was originally openable in QuPath using OpenSlide server, then the command could run.

If the image was originally openable( and only Bioformats can handle it because without this extension then QuPath cannot open it) in QuPath using BioFormats Extension, such as .vsi/.scn format, then the command raise error:

However, the image Leica-Fluorescence-1.scn could be hand-opened in QuPath, and after openning, I can run the same script by clicking inside QuPath without problem.

So the issue would be that how to let Bioformats extension usable when call QuPath from system command, but bot GUI.

Any suggestions are welcome.
================

20:56:55.735 [main] [INFO ] qupath.QuPath - Launching QuPath with args: -image, D:\\QMDownload\\6\\[failed_handOK]Leica_scn\\Leica-Fluorescence-1.scn, -script, D:\\QMDownload\\6\\Tiling_a_whole_slide_image.groovy 
20:56:55.829 [main] [ERROR] q.lib.images.servers.FileFormatInfo - Checking Big TIFF images currently not supported!!! 
Unsupported format or not found 
20:56:55.876 [main] [WARN ] q.i.i.servers.ImageJServerBuilder - Error opening D:\\QMDownload\\6\\[failed_handOK]Leica_scn\\Leica-Fluorescence-1.scn with ImageJ: Could not open D:\\QMDownload\\6\\[failed_handOK]Leica_scn\\Leica-Fluorescence-1.scn with ImageJ 
20:56:55.907 [main] [WARN ] q.l.i.servers.OpenslideServerBuilder - Unable to open D:\\QMDownload\\6\\[failed_handOK]Leica_scn\\Leica-Fluorescence-1.scn with OpenSlide: Can't find main image 
20:56:55.907 [main] [ERROR] q.l.i.servers.ImageServerProvider - Unable to build whole slide server - check your classpath for a suitable library (e.g. OpenSlide, BioFormats)
	 
20:56:55.907 [main] [ERROR] q.l.i.servers.ImageServerProvider - C:\Program Files\QuPath\app\QuPathApp.jar;qupath/qupath-core-0.1.2.jar;qupath/qupath-core-awt-0.1.2.jar;qupath/qupath-core-processing-0.1.2.jar;qupath/qupath-core-processing-awt-0.1.2.jar;qupath/qupath-extension-ij-0.1.2.jar;qupath/qupath-extension-input-0.1.2.jar;qupath/qupath-extension-opencv-0.1.2.jar;qupath/qupath-extension-openslide-0.1.2.jar;qupath/qupath-extension-pen-0.1.2.jar;qupath/qupath-extension-script-editor-0.1.2.jar;qupath/qupath-gui-fx-0.1.2.jar;qupath/qupath-processing-ij-0.1.2.jar;qupath/qupath-processing-opencv-0.1.2.jar;jars/commons-math3-3.6.1.jar;jars/controlsfx-8.40.12.jar;jars/flowless-0.4.5.jar;jars/groovy-2.4.7.jar;jars/groovy-jsr223-2.4.7.jar;jars/gson-2.8.0.jar;jars/ij-1.51g.jar;jars/jfxtras-common-8.0-r5.jar;jars/jfxtras-menu-8.0-r5.jar;jars/jinput-2.0.6.jar;jars/jpen-2-150301.jar;jars/jutils-1.0.0.jar;jars/logback-classic-1.1.7.jar;jars/logback-core-1.1.7.jar;jars/opencv-3.1.0.jar;jars/openslide-3.4.1_2.jar;jars/packager.jar;jars/reactfx-2.0-M4u1.jar;jars/richtextfx-0.6.10.jar;jars/slf4j-api-1.7.20.jar;jars/undofx-1.2.jar;jars/wellbehavedfx-0.1.1.jar 
20:56:56.627 [main] [ERROR] qupath.QuPath - Error running script! 
javax.script.ScriptException: javax.script.ScriptException: java.lang.NullPointerException 
	at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:155) 
	at qupath.QuPath.main(Unknown Source) 
Caused by: javax.script.ScriptException: java.lang.NullPointerException 
	at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:346) 
	at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:152) 
	... 1 common frames omitted 
Caused by: java.lang.NullPointerException: null 
	at qupath.imagej.images.servers.BufferedImagePlusServer.getPath(BufferedImagePlusServer.java:71) 
	at qupath.lib.images.servers.ImageServer$getPath.call(Unknown Source) 
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48) 
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113) 
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:117) 
	at Script1.run(Script1.groovy:42) 
	at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:343) 
	... 2 common frames omitted

**Comment by petebankhead** [comment]
Somehow you need the Bio-Formats extension and `bioformats_package.jar` added to the classpath from the command line, e.g. see https://docs.oracle.com/javase/8/docs/technotes/tools/windows/classpath.html

As I recall, the rest of the classpath entries are there because they are included inside `QuPathApp.jar` (in the manifest).

Eventually QuPath needs a better command line interface that takes care of loading extensions... but for now, I think/hope setting the classpath yourself should work.  I haven't tried it myself though.

**Comment by rmd06** [comment]
"C:\Program Files\QuPath\app\QuPathApp.jar" -image "D:\\QMDownload\\6\\[failed_handOK]Leica_scn\\Leica-Fluorescence-1.scn" -script "D:\\QMDownload\\6\\Tiling_a_whole_slide_image.groovy" -classpath "C:\\Users\\rmd\\QuPath\\extensions"

The error is still the same

**Comment by petebankhead** [comment]
Passing the directory isn't enough - you'd need to specify either the .jar files or use wild cards.  The link I posted has a section on using wild cards (I won't try to guess the exact right command as it is a while since I've had to use the classpath like this...).

**Comment by rmd06** [comment]
I have tried all the three forms:
  aClasspath = 'C:\Users\rmd\QuPath\extensions';
    aClasspath = 'C:\Users\rmd\QuPath\extensions\*';
    aClasspath = 'C:\Users\rmd\QuPath\extensions\qupath-extension-bioformats.jar;C:\Users\rmd\QuPath\extensions\bioformats_package.jar';

All failed.
It seems that the -classpath argument was ignored,. Indeel in the source I have seen that 

src/main/java/qupath/QuPath.java
=========
String SCRIPT_KEY = "-script";
String IMAGE_KEY = "-image";
 if (map.containsKey("-script")) {
...
==========
There was no -classpath option here.

**Comment by petebankhead** [comment]
The classpath would be handled before reaching QuPath, so it doesn't need appear at that point in the QuPath code.  Not sure what you mean with `aClasspath`; the syntax is more like this:
```
"C:\Program Files\QuPath\app\QuPathApp.jar" -image "D:\QMDownload\6\Leica-Fluorescence-1.scn" -script "D:\QMDownload\6\Tiling_a_whole_slide_image.groovy" -classpath "C:\Users\rmd\QuPath\extensions\qupath-extension-bioformats.jar;C:\Users\rmd\QuPath\extensions\bioformats_package.jar"
```
or possibly
```
"C:\Program Files\QuPath\app\QuPathApp.jar" -image "D:\QMDownload\6\Leica_scn\Leica-Fluorescence-1.scn" -script "D:\QMDownload\6\Tiling_a_whole_slide_image.groovy" -cp "C:\Program Files\QuPath\app\*;C:\Users\rmd\QuPath\extensions\*"
```
I'm not able to test these currently, but hopefully you can find a combination that works.

**Comment by rmd06** [comment]
Oh I just now found that you are the author of the great Image J book <Analyzing fluorescence microscopy images with ImageJ>. This book was the first image processing book I read and it indeed opens a new world for me.

**Comment by Svidro** [comment]
I think Pete's last post pretty much covers it, but you can also check the forum for command line script examples.

https://groups.google.com/forum/#!searchin/qupath-users/command$20line%7Csort:date/qupath-users/jGoPnfcFk6o/_UHKLdJYAgAJ

**Comment by petebankhead** [comment]
Keeping this apart from my current bonfire of issues as it warrants further investigation. Since Bio-Formats is now included by default I presume it should work in v0.2.0, but further extensions really should be picked up from the command line... and if they aren't, that would be a bug.

**Comment by rmd06** [comment]
Hi, Pete,
Yes, Bioformats worked when invoked from command line in v0.2.0 now.
